homeassistant/components/smartenergy_awattar/__init__.py

```diff
@@ -0,0 +1,228 @@
+
+MIN_UPDATE_INTERVAL: timedelta = timedelta(seconds=10)
+DEFAULT_UPDATE_INTERVAL: timedelta = timedelta(seconds=10)
+
+PLATFORMS: list[str] = [
+    SENSOR_DOMAIN,
+]
+
+# Configuration validation
+CONFIG_SCHEMA: vol.Schema = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_COUNTRY): vol.In(sorted(CONF_COUNTRY_LIST)),
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=DEFAULT_UPDATE_INTERVAL
+                ): vol.All(cv.time_period, vol.Clamp(min=MIN_UPDATE_INTERVAL)),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def ping_awattar(hass: HomeAssistant) -> None:
+    """Make a call to Awattar. If it fails raise an error."""
+
+    try:
+        api: AwattarApi = hass.data[DOMAIN][INIT_STATE][API]
+        await hass.async_add_executor_job(api.get_electricity_price)
+    except (aiohttp.ClientError, RuntimeError) as ex:
+        raise ConfigEntryNotReady(ex) from ex
+
+
+def _setup_coordinator(
+    hass: HomeAssistant,
+    scan_interval: timedelta,
+    coordinator_name: str,
+) -> DataUpdateCoordinator:
+    """Initialize the coordinator with empty state."""
+    _LOGGER.debug("Configuring coordinator=%s", coordinator_name)
+
+    state_fetcher: StateFetcher = StateFetcher(hass)
+    coordinator: DataUpdateCoordinator[dict] = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=state_fetcher.fetch_states,
+        update_interval=scan_interval,
+    )
+    state_fetcher.coordinator = coordinator
+    hass.data[DOMAIN][coordinator_name] = coordinator
+
+    return coordinator
+
+
+def _setup_api(config: ConfigType) -> dict:
+    """Initialize the API and save the reference in the state."""
+    awattar_api = {}
+
+    country: str = config[DOMAIN].get(CONF_COUNTRY)
+    _LOGGER.debug("Configuring Awattar API")
+    awattar_api = init_state(AWATTAR_API_URL[country])
+    _LOGGER.debug("Configured Awattar API")
+
+    return awattar_api
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up the Awattar defined via the UI.
+
+    - setup of the API
+    - coordinator
+    - sensors.
+    """
+    data = dict(config_entry.data)
+    entry_id = config_entry.entry_id
+
+    _LOGGER.debug(
+        "Setting up an entry with id=%s",
+        entry_id,
+    )
+
+    # scan interval is provided as an integer, but has to be an interval
+    scan_interval: timedelta = timedelta(seconds=data[CONF_SCAN_INTERVAL])
+    country: str = data[CONF_COUNTRY]
+
+    _LOGGER.debug("Configuring Awattar API")
+    hass.data[DOMAIN][INIT_STATE] = init_state(AWATTAR_API_URL[country])
+
+    # handle platform not ready
+    await ping_awattar(hass)
+
+    await _setup_coordinator(
+        hass,
+        scan_interval,
+        f"{entry_id}_coordinator",
+    ).async_config_entry_first_refresh()
+
+    hass.data[DOMAIN][entry_id] = data
```

Comment:
your hass.data is a bit complicated, you may want to look into using a model. Look at the lookin integration for an example

Reply:
Don't think it's that complicated, 2-3 key value pairs. Tried to introduce models, but imo it brought more complexity than clarity. Could be that I did also something incorrectly.